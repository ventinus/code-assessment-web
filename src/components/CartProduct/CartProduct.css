.cart-product + .cart-product {
  margin-top: 20px;
  border-top: 1px solid #d8d8d8;
  padding-top: 20px; }

.cart-product img {
  width: 110px; }

.cart-product__info {
  display: flex;
  flex-direction: row;
  align-items: start; }

.cart-product__info__details {
  margin-left: 15px;
  display: flex;
  flex-direction: column;
  align-items: start;
  align-self: stretch; }

.cart-product__info__details__remove-btn {
  color: #cc1d39;
  margin-top: auto;
  transition: opacity 0.3s ease-in-out 0s; }
  .cart-product__info__details__remove-btn:hover {
    opacity: 0.5; }

.cart-product__quantity {
  display: flex;
  flex-direction: row;
  align-items: stretch;
  margin-top: 20px;
  max-width: 300px; }
  .cart-product__quantity button {
    width: 95px;
    padding: 20px; }

.cart-product__quantity__value {
  flex-grow: 1;
  text-align: center;
  border-top: 1px solid #f5f5f5;
  border-bottom: 1px solid #f5f5f5;
  padding-top: 11px; }
