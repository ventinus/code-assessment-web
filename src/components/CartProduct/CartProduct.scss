@import '../../scss/variables/mixins.scss';
@import '../../scss/variables/variables.scss';

.cart-product {

  & + .cart-product {
    margin-top: 20px;
    border-top: 1px solid $grey-4;
    padding-top: 20px;
  }

  img { width: 110px; }
}

.cart-product__info {
  display: flex;
  flex-direction: row;
  align-items: start;
}

.cart-product__info__details {
  margin-left: 15px;
  display: flex;
  flex-direction: column;
  align-items: start;
  align-self: stretch;
}

.cart-product__info__details__remove-btn {
  color: $red-1;
  margin-top: auto;
  transition: default-transition(opacity);

  &:hover { opacity: 0.5; }
}

.cart-product__quantity {
  display: flex;
  flex-direction: row;
  align-items: stretch;
  margin-top: 20px;
  max-width: 300px;

  button {
    width: 95px;
    padding: 20px;
  }
}

.cart-product__quantity__value {
  flex-grow: 1;
  text-align: center;
  border-top: 1px solid $grey-2;
  border-bottom: 1px solid $grey-2;
  padding-top: 11px;
}
