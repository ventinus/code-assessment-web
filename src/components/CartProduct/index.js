import React from 'react'
import PropTypes from 'prop-types'

import {
  DecButton,
  IncButton,
} from '..'

import { toMoney } from '../../helpers'

import './CartProduct.css'

const CartProduct = ({
  img = {},
  id,
  title,
  price,
  currency,
  quantity,
  inventory,
  onAddToCartClicked,
  onRemoveFromCartClicked,
}) => (
  <li className="cart-product">
    <div className="cart-product__info">
      <img src={img.src} alt={img.alt} />
      <div className="cart-product__info__details">
        <h2 className="type--a7">{title}</h2>
        <p className="type--a8" style={{marginTop: 5}}>{toMoney(price, currency)}</p>
        <button
          className="cart-product__info__details__remove-btn type--a8"
          type="button"
          onClick={() => onRemoveFromCartClicked(id, quantity)}>
          Remove
        </button>
      </div>
    </div>
    <div className="cart-product__quantity">
      <DecButton onClick={() => onRemoveFromCartClicked(id)} >
        Remove one item from the cart
      </DecButton>
      <span className="cart-product__quantity__value type--a8">{quantity}</span>
      <IncButton onClick={() => onAddToCartClicked(id)} disabled={inventory <= 0} >
        Add one item to cart
      </IncButton>
    </div>
  </li>
)

CartProduct.propTypes = {
  img: PropTypes.shape({
    src: PropTypes.string,
    alt: PropTypes.string,
  }),
  id: PropTypes.number,
  title: PropTypes.string,
  price: PropTypes.number,
  currency: PropTypes.string,
  quantity: PropTypes.number,
  inventory: PropTypes.number,
  onAddToCartClicked: PropTypes.func,
  onRemoveFromCartClicked: PropTypes.func,
}

export default CartProduct
