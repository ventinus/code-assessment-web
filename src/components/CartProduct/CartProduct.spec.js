import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'
import CartProduct from '.'
import {
  DecButton,
  IncButton
} from '..'

const DEFAULT_PROPS = {
  id: 1,
  title: 'Cart Product',
  price: 10,
  quantity: 5,
  inventory: 2,
}

const setup = (props = {}) => {
  const actions = {
    onAddToCartClicked: jest.fn(),
    onRemoveFromCartClicked: jest.fn(),
  }

  const component = mount(
    <CartProduct
      {...Object.assign({}, DEFAULT_PROPS, props)}
      {...actions}
    />
  )

  return {
    actions,
    removeBtn: component.find('button').at(0),
    decBtn: component.find(DecButton),
    incBtn: component.find(IncButton),
  }
}

describe('CartProduct component', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(<CartProduct>children</CartProduct>)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should call onRemoveFromCartClicked with the id and quantity', () => {
    const { removeBtn, actions } = setup()
    removeBtn.simulate('click')
    expect(actions.onRemoveFromCartClicked).toHaveBeenCalledWith(1, 5)
  })

  it('should call onRemoveFromCartClicked with the id', () => {
    const { decBtn, actions } = setup()
    decBtn.simulate('click')
    expect(actions.onRemoveFromCartClicked).toHaveBeenCalledWith(1)
  })

  it('should call onAddToCartClicked with the id', () => {
    const { incBtn, actions } = setup()
    incBtn.simulate('click')
    expect(actions.onAddToCartClicked).toHaveBeenCalledWith(1)
  })

  describe('with no inventory', () => {
    it('should not call onAddToCartClicked', () => {
      const { incBtn, actions } = setup({inventory: 0})
      incBtn.simulate('click')
      expect(actions.onAddToCartClicked).not.toHaveBeenCalled()
    })
  })
})
