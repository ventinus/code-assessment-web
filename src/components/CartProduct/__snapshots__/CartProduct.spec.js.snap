// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartProduct component renders correctly 1`] = `
<li
  className="cart-product"
>
  <div
    className="cart-product__info"
  >
    <img />
    <div
      className="cart-product__info__details"
    >
      <h2
        className="type--a7"
      />
      <p
        className="type--a8"
        style={
          Object {
            "marginTop": 5,
          }
        }
      >
        $NaN
      </p>
      <button
        className="cart-product__info__details__remove-btn type--a8"
        onClick={[Function]}
        type="button"
      >
        Remove
      </button>
    </div>
  </div>
  <div
    className="cart-product__quantity"
  >
    <button
      className="button type--a8  button--secondary button--dec"
      disabled=""
      onClick={[Function]}
      type="button"
    >
      <span
        className="is-visually-hidden"
      >
        Remove one item from the cart
      </span>
    </button>
    <span
      className="cart-product__quantity__value type--a8"
    />
    <button
      className="button type--a8  button--secondary-alt button--inc"
      disabled=""
      onClick={[Function]}
      type="button"
    >
      <span
        className="is-visually-hidden"
      >
        Add one item to cart
      </span>
    </button>
  </div>
</li>
`;
